### ISSUE TITLE
YOLOv5 Output Size Issue

### ISSUE THREAD

**Issue by lllittleX** [ultralytics/yolov5#13008]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I want to convert the results to ONNX first, and then to RKNN. However, when testing the ONNX model, I found that my output sizes are as follows:
torch.Size([1, 15120, 133])
torch.Size([1, 3, 48, 80, 133])
torch.Size([1, 3, 24, 40, 133])
torch.Size([1, 3, 12, 20, 133])
But from what I understand, the usual size is 80,80 instead of 48,80. I would like to ask if this is normal?

**Comment by glenn-jocher** [comment]
@lllittleX hello! Thanks for reaching out with your question about the output size from the YOLOv5 model when converted to ONNX.

The output sizes you are seeing (e.g., `torch.Size([1, 3, 48, 80, 133])`) depend on the input image size you are using and the architecture of the model. Normally, YOLOv5 uses a grid that scales down from the input size by factors of 32, 16, and 8. If your input size isn't a multiple of these, the resulting grid sizes could appear unconventional, as is likely the case with a dimension of 80 instead of the usual 40 or 160, etc. 

Ensure that your input image size is a multiple of the largest factor (generally 32) to maintain consistent grid dimensions across the different layers. You can adjust this in your initial configuration before training or inference.
  
Let's verify that input sizes and model configurations align with the expected outputs. Good luck with your conversions! 🚀

**Comment by VishnuRao27** [comment]
Hi there!
I searched quite a lot for this but I couldnt really find it anywhere, let me give a little bit of context:
I am trying to perform knowledge distillation on yolov5n using yolov5x which are both already trained on my custom dataset containing 2 classes.
I saw this issue #13124 where I tutorial for knowledge distillation was given and I see that my pretrained custom_model.pt when loaded through pytorch (autoshape False) in eval mode gives 2 outputs:
1. A tensor of shape (batch_size, 25200, num_classes+5) ( [4, 25200, 7] in my case which is supposed to be [x, y, w, h, objectness_score, class_logits...] ? )
2. A list (of length 3) of tensors of shape (batch_size, 3, 80, 80, num_classes+5), (batch_size, 3, 40, 40, num_classes+5) and (batch_size, 3, 20, 20, num_classes+5). I dont know what the 7 values here correspond to.

The same model in train mode gives only the 2nd output.

At first I thought just stacking them would give me the same tensor as output1 but then I compared the values to see they are completely different even after applying softmax to them. 

Now for my problem:
Since knowledge distillation relies on the output probabilities of the teacher model which is in eval mode, I figured that the raw outputs from the teacher can be used directly since it gives the probabilities of each class. But since the student is in the train mode, which gives only output 2, how do I convert this list of tensors of shape (batch_size, 3, x, x, num_classes+5) to a raw logs score like how I see in output1 ? 
I do not obviously want to switch to eval mode because I want to use these values to compute the kl_div loss and backpropogate.

**Comment by VishnuRao27** [comment]
Hi there!
I searched quite a lot for this but I couldnt really find it anywhere, let me give a little bit of context:
I am trying to perform knowledge distillation on yolov5n using yolov5x which are both already trained on my custom dataset containing 2 classes.
I saw this issue #13124 where a tutorial for knowledge distillation was given and I see that my pretrained custom_model.pt when loaded through pytorch (autoshape False) in eval mode gives 2 outputs:
1. A tensor of shape (batch_size, 25200, num_classes+5) ( [4, 25200, 7] in my case which is supposed to be [x, y, w, h, objectness_score, class_logits...] ? )
2. A list (of length 3) of tensors of shape (batch_size, 3, 80, 80, num_classes+5), (batch_size, 3, 40, 40, num_classes+5) and (batch_size, 3, 20, 20, num_classes+5). I dont know what the 7 values here correspond to.

The same model in train mode gives only the 2nd output.

At first I thought just stacking them would give me the same tensor as output1 but then I compared the values to see they are completely different even after applying softmax to them. 
I experimented by actually applying softmax to the concatenated version of output 2, saving it and comparing it with the values of output1 saved without any modifications.

Now for my problem:
Since knowledge distillation relies on the output probabilities of the teacher model which is in eval mode, I figured that the raw outputs from the teacher can be used directly since it gives the probabilities of each class. But since the student is in the train mode, which gives only output 2, how do I convert this list of tensors of shape (batch_size, 3, x, x, num_classes+5) to a raw logs score like how I see in output1 ? 
I do not obviously want to switch to eval mode because I want to use these values to compute the kl_div loss and backpropogate.
I can also just stack the output2 of student and teacher to give something with the shape of output1 and pass it to kl_div but then if the values are not actually the probability distributions of each class then it defeats the purpose of knowledge distillation.

**Comment by pderrenger** [comment]
@VishnuRao27 for knowledge distillation in YOLOv5, ensure both teacher and student models are in **train mode** to access raw feature outputs (output 2). The three tensors in the list correspond to detection layers at different scales. To align outputs:

1. **Reshape** each detection layer tensor to `[batch, anchors, (xywh + obj + classes)]`
2. **Concatenate** along the anchors dimension to match the `25200` anchors in eval output
3. **Apply sigmoid** to class logits (not softmax) as YOLOv5 uses independent class probabilities

Example reshaping:
```python
student_output = torch.cat([x.view(x.shape[0], -1, x.shape[-1]) for x in train_output], 1)
teacher_output = torch.cat([x.view(x.shape[0], -1, x.shape[-1]) for x in teacher_train_output], 1)
```

For implementation details, see the YOLOv5 `Model` class in models/yolo.py which handles output reshaping internally during training.

**Comment by VishnuRao27** [comment]
Hi @pderrenger ! Thank you for your reply
This surely helps, I was doing it right but applying the wrong function in the end (softmax instead of sigmoid).
I just have one question, I have the teacher model currently in eval mode and I am doing this:


```
teacher.eval()
with torch.no_grad():
     _, teacher_fm = teacher(images)
```

This is also the output of the 3 detection layers not concatenated yet right?

Edit: So if I follow the steps, use that code snippet that you shared above and get the outputs and apply sigmoid to them, I will basically get the output1 that is returned along with output 2 in the inference mode right?

**Comment by pderrenger** [comment]
Yes, when using `teacher.eval()`, the `teacher_fm` output will be the raw detection layers (unconcatenated). Concatenating them via `torch.cat([x.view(...) for x in teacher_fm], 1)` and applying sigmoid to class/objectness outputs will align with the `output1` format. The Model class in models/yolo.py handles this reshaping internally during inference.
